### PR TITLE
ci: migrate codebase to use NX (using lerna integration)

### PR DESCRIPTION
**Pull request by jakemhiller**
### Overview

Switch to NX for task caching and package builds, further integration in future updates

Next steps

- [ ] convert repo to use nx test utilities & cache in CI
- [ ] convert repo to use nx lint utilities & cache in CI

### PR Checklist

- [ ] Related to designs:
- [ ] Related to JIRA ticket: ABC-123
- [ ] I have run this code to verify it works
- [ ] This PR includes unit tests for the code change
- [ ] This PR includes testing instructions tests for the code change
- [ ] The alpha package of this PR is passing end-to-end tests in all relevant Codecademy repositories

#### Testing Instructions

To test the dev experience, pull down this branch and run `yarn build`, then run it again. The second time everything should be cached and the build should be basically instant.

Now change a package that is shared by a few other packages, like `@codecademy/gamut-illustrations`, and run `yarn build` again. You should see `gamut`, `gamut-labs`, and `monaco` are rebuilt, and all other builds are cached.

Everything should be the same for the builds, i didn't modify how the packages were being built, just what was building them. you can confirm by comparing the contents of the latest version of the packages with the alpha versions from this PR.

https://unpkg.com/browse/@codecademy/gamut-icons@7.4.3-alpha.7b7eab.0/dist/icons/regular/AccountingCoinsIcon.js
https://unpkg.com/browse/@codecademy/gamut-icons@7.4.2/dist/icons/regular/AccountingCoinsIcon.js

#### PR Links and Envs

| Repository   | PR Link                                                  | PR Env                                                   |
| :----------- | :------------------------------------------------------- | :------------------------------------------------------- |
| Monolith     | [Monolith PR](https://github.com/codecademy-engineering/Codecademy/pull/31722)  | [Monolith Env](http://www.google.fr/ 'Named link title') |

### CODE REVIEW

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>eslint-plugin-gamut@2.0.1-alpha.c246a8.0<br/>@codecademy/gamut-icons@7.4.3-alpha.c246a8.0<br/>@codecademy/gamut-illustrations@0.28.3-alpha.c246a8.0<br/>@codecademy/gamut-kit@0.6.164-alpha.c246a8.0<br/>@codecademy/gamut-labs@32.7.7-alpha.c246a8.0<br/>@codecademy/gamut-patterns@0.8.9-alpha.c246a8.0<br/>@codecademy/gamut-styles@13.4.1-alpha.c246a8.0<br/>@codecademy/gamut-tests@4.0.8-alpha.c246a8.0<br/>@codecademy/gamut@48.3.0-alpha.c246a8.0<br/>@codecademy/macros@3.0.1-alpha.c246a8.0<br/>@codecademy/markdown-overrides@0.9.6-alpha.c246a8.0<br/>@codecademy/monaco@0.6.59-alpha.c246a8.0<br/>@codecademy/styleguide@58.6.0-alpha.c246a8.0<br/>@codecademy/variance@0.20.1-alpha.c246a8.0<br/></pre>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>eslint-plugin-gamut@2.0.1-alpha.b947b1.0<br/>@codecademy/gamut-icons@7.4.3-alpha.b947b1.0<br/>@codecademy/gamut-illustrations@0.28.3-alpha.b947b1.0<br/>@codecademy/gamut-kit@0.6.164-alpha.b947b1.0<br/>@codecademy/gamut-labs@32.7.7-alpha.b947b1.0<br/>@codecademy/gamut-patterns@0.8.9-alpha.b947b1.0<br/>@codecademy/gamut-styles@13.4.1-alpha.b947b1.0<br/>@codecademy/gamut-tests@4.0.8-alpha.b947b1.0<br/>@codecademy/gamut@48.3.0-alpha.b947b1.0<br/>@codecademy/macros@3.0.1-alpha.b947b1.0<br/>@codecademy/markdown-overrides@0.9.6-alpha.b947b1.0<br/>@codecademy/monaco@0.6.59-alpha.b947b1.0<br/>@codecademy/styleguide@58.6.0-alpha.b947b1.0<br/>@codecademy/variance@0.20.1-alpha.b947b1.0<br/></pre>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>eslint-plugin-gamut@2.0.1-alpha.ccf430.0<br/>@codecademy/gamut-icons@7.4.3-alpha.ccf430.0<br/>@codecademy/gamut-illustrations@0.28.3-alpha.ccf430.0<br/>@codecademy/gamut-kit@0.6.164-alpha.ccf430.0<br/>@codecademy/gamut-labs@32.7.7-alpha.ccf430.0<br/>@codecademy/gamut-patterns@0.8.9-alpha.ccf430.0<br/>@codecademy/gamut-styles@13.4.1-alpha.ccf430.0<br/>@codecademy/gamut-tests@4.0.8-alpha.ccf430.0<br/>@codecademy/gamut@48.3.0-alpha.ccf430.0<br/>@codecademy/macros@3.0.1-alpha.ccf430.0<br/>@codecademy/markdown-overrides@0.9.6-alpha.ccf430.0<br/>@codecademy/monaco@0.6.59-alpha.ccf430.0<br/>@codecademy/styleguide@58.6.0-alpha.ccf430.0<br/>@codecademy/variance@0.20.1-alpha.ccf430.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://62d865afe673770de3c84bec--gamut-preview.netlify.app>https://62d865afe673770de3c84bec--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/62d865afe673770de3c84bec>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>eslint-plugin-gamut@2.0.1-alpha.7b7eab.0<br/>@codecademy/gamut-icons@7.4.3-alpha.7b7eab.0<br/>@codecademy/gamut-illustrations@0.28.3-alpha.7b7eab.0<br/>@codecademy/gamut-kit@0.6.165-alpha.7b7eab.0<br/>@codecademy/gamut-labs@33.0.1-alpha.7b7eab.0<br/>@codecademy/gamut-patterns@0.8.9-alpha.7b7eab.0<br/>@codecademy/gamut-styles@13.4.1-alpha.7b7eab.0<br/>@codecademy/gamut-tests@4.0.8-alpha.7b7eab.0<br/>@codecademy/gamut@48.3.1-alpha.7b7eab.0<br/>@codecademy/macros@3.0.1-alpha.7b7eab.0<br/>@codecademy/markdown-overrides@0.9.6-alpha.7b7eab.0<br/>@codecademy/monaco@0.6.60-alpha.7b7eab.0<br/>@codecademy/styleguide@59.0.1-alpha.7b7eab.0<br/>@codecademy/variance@0.20.1-alpha.7b7eab.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://62d9fcfd67db2b24a5a50a79--gamut-preview.netlify.app>https://62d9fcfd67db2b24a5a50a79--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/62d9fcfd67db2b24a5a50a79>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>eslint-plugin-gamut@2.0.1-alpha.6c28e2.0<br/>@codecademy/gamut-icons@7.4.3-alpha.6c28e2.0<br/>@codecademy/gamut-illustrations@0.28.3-alpha.6c28e2.0<br/>@codecademy/gamut-kit@0.6.165-alpha.6c28e2.0<br/>@codecademy/gamut-labs@33.0.1-alpha.6c28e2.0<br/>@codecademy/gamut-patterns@0.8.9-alpha.6c28e2.0<br/>@codecademy/gamut-styles@13.4.1-alpha.6c28e2.0<br/>@codecademy/gamut-tests@4.0.8-alpha.6c28e2.0<br/>@codecademy/gamut@48.3.1-alpha.6c28e2.0<br/>@codecademy/macros@3.0.1-alpha.6c28e2.0<br/>@codecademy/markdown-overrides@0.9.6-alpha.6c28e2.0<br/>@codecademy/monaco@0.6.60-alpha.6c28e2.0<br/>@codecademy/styleguide@59.0.1-alpha.6c28e2.0<br/>@codecademy/variance@0.20.1-alpha.6c28e2.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://62dac0ed8aec3806196f0e9f--gamut-preview.netlify.app>https://62dac0ed8aec3806196f0e9f--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/62dac0ed8aec3806196f0e9f>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>eslint-plugin-gamut@2.0.1-alpha.9bfc43.0<br/>@codecademy/gamut-icons@7.4.3-alpha.9bfc43.0<br/>@codecademy/gamut-illustrations@0.28.3-alpha.9bfc43.0<br/>@codecademy/gamut-kit@0.6.165-alpha.9bfc43.0<br/>@codecademy/gamut-labs@33.0.1-alpha.9bfc43.0<br/>@codecademy/gamut-patterns@0.8.9-alpha.9bfc43.0<br/>@codecademy/gamut-styles@13.4.1-alpha.9bfc43.0<br/>@codecademy/gamut-tests@4.0.8-alpha.9bfc43.0<br/>@codecademy/gamut@48.3.1-alpha.9bfc43.0<br/>@codecademy/macros@3.0.1-alpha.9bfc43.0<br/>@codecademy/markdown-overrides@0.9.6-alpha.9bfc43.0<br/>@codecademy/monaco@0.6.60-alpha.9bfc43.0<br/>@codecademy/styleguide@59.0.1-alpha.9bfc43.0<br/>@codecademy/variance@0.20.1-alpha.9bfc43.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://62dae0870571081f39933c43--gamut-preview.netlify.app>https://62dae0870571081f39933c43--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/62dae0870571081f39933c43>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>eslint-plugin-gamut@2.0.1-alpha.825c81.0<br/>@codecademy/gamut-icons@7.4.3-alpha.825c81.0<br/>@codecademy/gamut-illustrations@0.28.3-alpha.825c81.0<br/>@codecademy/gamut-kit@0.6.165-alpha.825c81.0<br/>@codecademy/gamut-labs@33.0.1-alpha.825c81.0<br/>@codecademy/gamut-patterns@0.8.9-alpha.825c81.0<br/>@codecademy/gamut-styles@13.4.1-alpha.825c81.0<br/>@codecademy/gamut-tests@4.0.8-alpha.825c81.0<br/>@codecademy/gamut@48.3.1-alpha.825c81.0<br/>@codecademy/macros@3.0.1-alpha.825c81.0<br/>@codecademy/markdown-overrides@0.9.6-alpha.825c81.0<br/>@codecademy/monaco@0.6.60-alpha.825c81.0<br/>@codecademy/styleguide@59.0.1-alpha.825c81.0<br/>@codecademy/variance@0.20.1-alpha.825c81.0<br/></pre>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>eslint-plugin-gamut@2.0.1-alpha.f24332.0<br/>@codecademy/gamut-icons@7.4.3-alpha.f24332.0<br/>@codecademy/gamut-illustrations@0.28.3-alpha.f24332.0<br/>@codecademy/gamut-kit@0.6.165-alpha.f24332.0<br/>@codecademy/gamut-labs@33.0.1-alpha.f24332.0<br/>@codecademy/gamut-patterns@0.8.9-alpha.f24332.0<br/>@codecademy/gamut-styles@13.4.1-alpha.f24332.0<br/>@codecademy/gamut-tests@4.0.8-alpha.f24332.0<br/>@codecademy/gamut@48.3.1-alpha.f24332.0<br/>@codecademy/macros@3.0.1-alpha.f24332.0<br/>@codecademy/markdown-overrides@0.9.6-alpha.f24332.0<br/>@codecademy/monaco@0.6.60-alpha.f24332.0<br/>@codecademy/styleguide@59.0.1-alpha.f24332.0<br/>@codecademy/variance@0.20.1-alpha.f24332.0<br/></pre>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>eslint-plugin-gamut@2.0.1-alpha.25d8b9.0<br/>@codecademy/gamut-icons@7.4.3-alpha.25d8b9.0<br/>@codecademy/gamut-illustrations@0.28.3-alpha.25d8b9.0<br/>@codecademy/gamut-kit@0.6.165-alpha.25d8b9.0<br/>@codecademy/gamut-labs@33.0.1-alpha.25d8b9.0<br/>@codecademy/gamut-patterns@0.8.9-alpha.25d8b9.0<br/>@codecademy/gamut-styles@13.4.1-alpha.25d8b9.0<br/>@codecademy/gamut-tests@4.0.8-alpha.25d8b9.0<br/>@codecademy/gamut@48.3.1-alpha.25d8b9.0<br/>@codecademy/macros@3.0.1-alpha.25d8b9.0<br/>@codecademy/markdown-overrides@0.9.6-alpha.25d8b9.0<br/>@codecademy/monaco@0.6.60-alpha.25d8b9.0<br/>@codecademy/styleguide@59.0.1-alpha.25d8b9.0<br/>@codecademy/variance@0.20.1-alpha.25d8b9.0<br/></pre>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>eslint-plugin-gamut@2.0.1-alpha.868d87.0<br/>@codecademy/gamut-icons@7.4.3-alpha.868d87.0<br/>@codecademy/gamut-illustrations@0.28.3-alpha.868d87.0<br/>@codecademy/gamut-kit@0.6.165-alpha.868d87.0<br/>@codecademy/gamut-labs@33.0.1-alpha.868d87.0<br/>@codecademy/gamut-patterns@0.8.9-alpha.868d87.0<br/>@codecademy/gamut-styles@13.4.1-alpha.868d87.0<br/>@codecademy/gamut-tests@4.0.8-alpha.868d87.0<br/>@codecademy/gamut@48.3.1-alpha.868d87.0<br/>@codecademy/macros@3.0.1-alpha.868d87.0<br/>@codecademy/markdown-overrides@0.9.6-alpha.868d87.0<br/>@codecademy/monaco@0.6.60-alpha.868d87.0<br/>@codecademy/styleguide@59.0.1-alpha.868d87.0<br/>@codecademy/variance@0.20.1-alpha.868d87.0<br/></pre>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>eslint-plugin-gamut@2.0.1-alpha.54c5f5.0<br/>@codecademy/gamut-icons@7.4.3-alpha.54c5f5.0<br/>@codecademy/gamut-illustrations@0.28.3-alpha.54c5f5.0<br/>@codecademy/gamut-kit@0.6.165-alpha.54c5f5.0<br/>@codecademy/gamut-labs@33.0.1-alpha.54c5f5.0<br/>@codecademy/gamut-patterns@0.8.9-alpha.54c5f5.0<br/>@codecademy/gamut-styles@13.4.1-alpha.54c5f5.0<br/>@codecademy/gamut-tests@4.0.8-alpha.54c5f5.0<br/>@codecademy/gamut@48.3.1-alpha.54c5f5.0<br/>@codecademy/macros@3.0.1-alpha.54c5f5.0<br/>@codecademy/markdown-overrides@0.9.6-alpha.54c5f5.0<br/>@codecademy/monaco@0.6.60-alpha.54c5f5.0<br/>@codecademy/styleguide@59.0.1-alpha.54c5f5.0<br/>@codecademy/variance@0.20.1-alpha.54c5f5.0<br/></pre>

File: nx.json [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      "style": "@emotion/styled",
+      "linter": "eslint"
+    },
+    "@nrwl/react": {
+      "application": {
+        "babel": true
+      }
+    }
+  }
+}
```
Comment (BandanaKM) [inline nx.json:51]: Is this file generated or did you intentionally set any properties here?

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>eslint-plugin-gamut@2.0.1-alpha.b9b96b.0<br/>@codecademy/gamut-icons@7.4.3-alpha.b9b96b.0<br/>@codecademy/gamut-illustrations@0.28.3-alpha.b9b96b.0<br/>@codecademy/gamut-kit@0.6.165-alpha.b9b96b.0<br/>@codecademy/gamut-labs@33.0.1-alpha.b9b96b.0<br/>@codecademy/gamut-patterns@0.8.9-alpha.b9b96b.0<br/>@codecademy/gamut-styles@13.4.1-alpha.b9b96b.0<br/>@codecademy/gamut-tests@4.0.8-alpha.b9b96b.0<br/>@codecademy/gamut@48.3.1-alpha.b9b96b.0<br/>@codecademy/macros@3.0.1-alpha.b9b96b.0<br/>@codecademy/markdown-overrides@0.9.6-alpha.b9b96b.0<br/>@codecademy/monaco@0.6.60-alpha.b9b96b.0<br/>@codecademy/styleguide@59.0.1-alpha.b9b96b.0<br/>@codecademy/variance@0.20.1-alpha.b9b96b.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://62db4890509fb631a7dcc6d0--gamut-preview.netlify.app>https://62db4890509fb631a7dcc6d0--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/62db4890509fb631a7dcc6d0>Deploy Logs</a></p>

File: nx.json [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        "cacheableOperations": ["build", "test", "verify", "build-storybook"],
+        "parallel": 3
+      }
+    }
+  },
+  "targetDefaults": {
+    "build": {
+      "dependsOn": ["^build"]
+    }
+  },
```
Comment (jakemhiller) [inline nx.json:32]: @BandanaKM this is mostly what i edited from the generated content of this file. I also messed around with the `generators` settings, but we're not really using the generators in this case.

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>eslint-plugin-gamut@2.0.1-alpha.fe86ea.0<br/>@codecademy/gamut-icons@7.4.3-alpha.fe86ea.0<br/>@codecademy/gamut-illustrations@0.28.3-alpha.fe86ea.0<br/>@codecademy/gamut-kit@0.6.165-alpha.fe86ea.0<br/>@codecademy/gamut-labs@33.0.1-alpha.fe86ea.0<br/>@codecademy/gamut-patterns@0.8.9-alpha.fe86ea.0<br/>@codecademy/gamut-styles@13.4.1-alpha.fe86ea.0<br/>@codecademy/gamut-tests@4.0.8-alpha.fe86ea.0<br/>@codecademy/gamut@48.3.1-alpha.fe86ea.0<br/>@codecademy/macros@3.0.1-alpha.fe86ea.0<br/>@codecademy/markdown-overrides@0.9.6-alpha.fe86ea.0<br/>@codecademy/monaco@0.6.60-alpha.fe86ea.0<br/>@codecademy/styleguide@59.0.1-alpha.fe86ea.0<br/>@codecademy/variance@0.20.1-alpha.fe86ea.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://62deb23cc82bc23717dd28db--gamut-preview.netlify.app>https://62deb23cc82bc23717dd28db--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/62deb23cc82bc23717dd28db>Deploy Logs</a></p>

File: README.md [hunk 3]
```diff
@@ -202,13 +202,13 @@ Notes:
...
+**NX**
 
-This monorepo uses [Turborepo](https://turborepo.org/) to cache previous builds locally and in CI.
+This monorepo uses [NX](https://nx.dev/) to cache previous builds locally and in CI.
 
-The config for Turborepo is located at [/turbo.json](/turbo.json).
+The config for NX is located at [/nx.json](/nx.json), along with `project.json` files for each package.
 
-To use Turborepo without extra configuration, if your package needs to be compiled, it should have a task called `build` that compiles it's files and puts them into a directory called `dist` inside the package folder. If you need a more complicated setup, you can read the docs and customize the configuration in `turbo.json`.
+For new packages, please use an NX generator plugin to create your initial package, this will ensure that all of the configuration for linting & testing is set up correctly.
```
Comment (BandanaKM) [inline README.md:211]: thinking we can either link to the plugin documentation or mention the generators in nx.json for folks new to nx

File: nx.json [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  "cli": {
+    "defaultCollection": "@nrwl/react"
+  },
+  "generators": {
+    "@nrwl/js:library": {
+      "buildable": true
```
Comment (BandanaKM) [inline nx.json:38]: any reason to leave our publishable and the linter here?

File: nx.json (hunk 2 above)
Comment (BandanaKM) [reply nx.json]: got it!

File: package.json [hunk 5]
```diff
@@ -89,25 +86,32 @@
...
+    "lodash": "^4.17.5",
     "mutationobserver-shim": "^0.3.3",
+    "nx": "14.4.3",
     "prettier": "2.2.1",
     "react": "^17.0.2",
     "react-dom": "^17.0.2",
     "react-helmet": "6.1.0",
     "react-test-renderer": "^17.0.2",
-    "turbo": "^1.1.2",
+    "svgo": "^1.3.2",
```
Comment (BandanaKM) [inline package.json:114]: oh, weird we didn't have this before

**Review by BandanaKM — APPROVED** [review]
really cool; maybe worth running by cass too, but looks pretty neat!

File: package.json (hunk 5 above)
Comment (jakemhiller) [reply package.json:114]: yeah, i moved a few of the dev dependencies up into the main package.json to simplify the move. I ended up moving some back but once we integrate [syncpack](https://github.com/JamieMason/syncpack) we should at least guarantee that everything stays at the same version.

File: nx.json (hunk 4 above)
Comment (jakemhiller) [reply nx.json:38]: not really, i think i'm just going to remove these defaults completely though, until we know the settings we're actually going to use.

**Pull request merged**